README.md
# apiks (aka API on EKS)

API-driven app with [PostgreSQL on RDS](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_PostgreSQL.html) as backend, flask &amp; kong as API gw &amp; Nginx as ingress controller

## Requirements

* `terraform` - for cluster provisioning
* `jq` (optional) - for cluster provisioning


CHANGELOG.md
# 0.0.5

1611495212 Dockerfile
1611522800 Docker build and push to the AWS ECR

# 0.0.4

* Flask refactoring and debug
* kong initial app

# 0.0.3

## Features

1610649215 flask api app
1610652032 Swagger updated

# 0.0.2

## Features

1610355538 Backend boilerplate
1610355640 Nginx reverse proxy boilerplate
1610398548 Provisioning - initial commit

# 0.0.1

## Documentation and Architecture

* Documentation boilerplate
* Project deployment diagram-as-a-code

## Features

* [PetClinic](https://github.com/spring-projects/spring-petclinic) SQL boilerplate & ER-diagram-as-a-code
1610285436 Initial commit
1610285723 Documentation
1610286055 PetClinic boilerplate
1610286131 PetClinic boilerplate
1610286719 project diagram as a code
1610291084 Sample schemas
1610291403 Sample Schemas
1610294255 diagram
1610308171 ER-diagram of PetClinic project
1610310774 SQL schema & data for PetClinic project
1610311471 Diagrams-as-a-code - Table Of Contents
1610311562 Project Documentation
1610311953 Readme
1610314028 tag:0.0.1
1610355538 Backend boilerplate
1610355640 Nginx reverse proxy boilerplate
1610398548 Provisioning - initial commit
1610398596 Add terraform to `.gitignore`
1610398860 0.0.2
1610649215 flask api app
1610652032 Swagger updated
1610652599 tag 0.0.3
1610990461 kong
1611431083 Flask refactoring and debug
1611431572 0.0.4
1611495212 Dockerfile
1611522576 Docker build and push to the AWS ECR
1611431572 0.0.4
1611495212 Dockerfile
1611522576 Docker build and push to the AWS ECR


backend/flask/api.py
from apiks_flask_API import *

# route to get all movies
@app.route('/pet', methods=['GET'])
def get_pets():
    '''Function to get all the movies in the database'''
    return jsonify({'pet': pet.get_all_pets()})

# route to get movie by id
@app.route('/pet/<int:id>', methods=['GET'])
def get_pet_by_id(id):
    return_value = pet.get_pet(id)
    return jsonify(return_value)

# route to add new movie
@app.route('/pet', methods=['POST'])
def add_pet():
    '''Function to add new movie to our database'''
    request_data = request.get_json()  # getting data from client
    pet.add_pet(request_data["pet_type_id"], request_data["name"],
                    request_data["client_id"], request_data["age"])
    response = Response("Pet added", 201, mimetype='application/json')
    return response

# route to delete movie using the DELETE method
@app.route('/pets/<int:id>', methods=['DELETE'])
def remove_pet(id):
    '''Function to delete movie from our database'''
    pet.delete_pet(id)
    response = Response("Movie Deleted", status=200, mimetype='application/json')
    return response

if __name__ == "__main__":
    app.run(port=1234, debug=True)


backend/flask/settings.py
#!/usr/bin/env python3

from flask import Flask, request, Response, jsonify
from flask_sqlalchemy import SQLAlchemy
import json
from flasgger import Swagger

app = Flask(__name__)
db = SQLAlchemy()
# Wrap the Api with swagger.docs. It is a thin wrapper around the Api class that adds some swagger smarts
# api = swagger.docs(Api(app), apiVersion='0.1')
swagger = Swagger(app)

app.config['SQLALCHEMY_DATABASE_URI'] = 'apiks.c8kflxch1q8h.eu-west-2.rds.amazonaws.com'
app.config['SQLALCHEMY_TRACK_MODFICATIONS'] = False

@app.route('/')
def hello():
    return "Hello World!"

@app.route("/session_list", methods=['GET'])
def session_list():
    return jsonify({'Session_list': SessionList.get_all_session_lists()})

class SessionList(db.Model):
    __tablename__ = 'session_list'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    client_id = db.Column(db.Integer, nullable=False)
    # nullable is false so the column can't be empty
    clinic_id = db.Column(db.Integer, nullable=False)
    doctor_id = db.Column(db.Integer, nullable=False)
    price_list_id = db.Column(db.Integer, nullable=False)

    def __repr__(self,db):
        return 'USER %r' % self.db

    def json(self):
        return {'id': self.id, 'client_id': self.clinic_id,
                'clinic_id': self.clinic_id, 'doctor_id': self.doctor_id, 'date_of_visit': self.date_of_visit, 'price_list_id': self.price_list_id}

    def add_session_list(_client_id, _clinic_id, _doctor_id, _date_of_visit, _price_list_id):
        new_session_list = session_list(client_id=_client_id, clinic_id=_clinic_id, doctor_id=_doctor_id, date_of_visit=_date_of_visit, price_list_id=_price_list_id)
        db.session.add(new_session_list)
        db.session.commit()  # commit changes to session

class PriceList(db.Model):
    __tablename__ = 'price_list'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    service = db.Column(db.String(30), nullable=False)
    # nullable is false so the column can't be empty
    price = db.Column(db.Integer, nullable=False)
    doctor_position_id = db.Column(db.Integer, nullable=False)

    def json(self):
        return {'id': self.id, 'service': self.service,
                'price': self.price, 'doctor_position_id': self.doctor_position_id}

class PetMedicalCard(db.Model):
    __tablename__ = 'pet_medical_card'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    doctor_id = db.Column(db.Integer, nullable=False)
    # nullable is false so the column can't be empty
    pet_id = db.Column(db.Integer, nullable=False)
    info = db.Column(db.String(1000), nullable=False)

    def json(self):
        return {'id': self.id, 'doctor_id': self.doctor_id,
                'pet_id': self.pet_id, 'info': self.info, 'date_of_visit': self.date_of_visit}

class Doctor(db.Model):
    __tablename__ = 'doctor'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    name = db.Column(db.String(30), nullable=False)
    # nullable is false so the column can't be empty
    doctor_position_id = db.Column(db.Integer, nullable=False)
    phone = db.Column(db.String(13), nullable=False)
    clinic_id = db.Column(db.Integer, nullable=False)
    cabinet = db.Column(db.Integer, nullable=False)

    def json(self):
        return {'id': self.id, 'name': self.name,
                'doctor_position_id': self.doctor_position_id, 'phone': self.phone, 'clinic_id': self.clinic_id, 'cabinet': self.cabinet}

class Clinic(db.Model):
    __tablename__ = 'clinic'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    address = db.Column(db.String(50), nullable=False)
    # nullable is false so the column can't be empty
    schedule_clinic = db.Column(db.String(20), nullable=False)
    phone = db.Column(db.String(20), nullable=False)

    def json(self):
        return {'id': self.id, 'address': self.address,
                'schedule_clinic': self.schedule_clinic, 'phone': self.phone}

class DoctorPosition(db.Model):
    __tablename__ = 'doctor_position'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    position = db.Column(db.String(20), nullable=False)
    # nullable is false so the column can't be empty

    def json(self):
        return {'id': self.id, 'position': self.position,}

class PetTable(db.Model):
    __tablename__ = 'Pet'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    pet_type_id = db.Column(db.Integer, nullable=False)
    # nullable is false so the column can't be empty
    name = db.Column(db.String(30), nullable=False)
    client_id = db.Column(db.Integer, nullable=False)
    age = db.Column(db.Integer, nullable=False)

    def json(self):
        return {'id': self.id, 'pet_type_id': self.pet_type_id,
                'name': self.name, 'client_id': self.client_id, 'age': self.age}

    def add_pet(_pet_type_id, _name, _client_id, _age):
        request_data = request.get_json()  # getting data from client
        pet.add_pet(request_data["pet_type_id"], request_data["name"],
            request_data["client_id"], request_data["age"])
        response = Response("Pet added", 201, mimetype='application/json')
        new_pet = PetTable(pet_type_id=_pet_type_id, name=_name, client_id=_client_id, age=_age)
        db.session.commit()  # commit changes to session
        return response

    def get_all_pets():
        return [PetTable.json(pet) for pet in PetTable.query.all()]

    def get_pet(_name):
        return [PetTable.json(PetTable.query.filter_by(name=_name).first())]

class Client(db.Model):
    __tablename__ = 'client'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    name = db.Column(db.String(30), nullable=False)
    # nullable is false so the column can't be empty
    phone = db.Column(db.String(13), nullable=False)
    email = db.Column(db.String(30), nullable=False)
    user_role_id = db.Column(db.Integer, nullable=False)

    def json(self):
        return {'id': self.id, 'name': self.name,
                'phone': self.phone, 'email': self.email, 'user_role': self.user_role}

class UserRole(db.Model):
    __tablename__ = 'user_role'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    role = db.Column(db.String(15), nullable=False)
    # nullable is false so the column can't be empty

    def json(self):
        return {'id': self.id, 'role': self.role}

class PetType(db.Model):
    __tablename__ = 'pet_type'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    type = db.Column(db.String(30), nullable=False)
    # nullable is false so the column can't be empty

    def json(self):
        return {'id': self.id, 'type': self.type}

#@app.route('/pet', methods=['GET'])
#def get_pets():
#    return jsonify({'pet': PetTable.get_all_pets()})

@app.route('/pet/<int:id>', methods=['GET'])
def get_pet_by_id(id):
    """Example endpoint returning a pet by it's id
    This is using docstrings for specifications.
    ---
    parameters:
      - name: Pet't id
        in: path
        type: string
        required: true
    responses:
      200:
        description: Information about the pet
    """
    return_value = PetTable.get_pet(id)
    return jsonify(return_value)

@app.route('/pet', methods=['POST'])
def add_pet():
    """Example endpoint adding a new pet to the list of pets
    This is using docstrings for specifications.
    ---
    parameters:
      - name: Pet's id
        in: path
        type: Integer
        required: true
        description: Type down the id of new pet
      - name: name
        in: path
        type: string
        required: true
        description: Name of the pet required
      - name: client id
        in: path
        type: Integer
        required: true
        description: Client id
      - name: age
        in: path
        type: Integer
        required: true
        description: Age of the pet
    responses:
      200:
        description: New pet was added
    """
    request_data = request.get_json()  # getting data from client
    pet.add_pet(request_data["pet_type_id"], request_data["name"],
                    request_data["client_id"], request_data["age"])
    response = Response("Pet added", 201, mimetype='application/json')
    return response

@app.route('/pets/<int:id>', methods=['DELETE'])
def remove_pet(id):
    """Example endpoint deleting a pet by it's id
    This is using docstrings for specifications.
    ---
    parameters:
      - name: Delete pet by it's id
        in: path
        type: string
        description: pet's id
        required: true
        default: all
    responses:
      200:
        description: Pet was deleted
    """
    pet.delete_pet(id)
    response = Response("Pet Deleted", status=200, mimetype='application/json')
    return response

@app.route('/pet', methods=['GET'])
def get_pets():
    """Example endpoint returning a list of pets
    This is using docstrings for specifications.
    ---
    parameters:
      - name: Pets
        in: path
        type: string
        enum: ['all', 'cats', 'dogs']
        required: true
        default: all
    definitions:
      Pets:
        type: object
        properties:
          pet_name:
            type: array
            items:
              $ref: '#/pet'
      Pet:
        type: string
    responses:
      200:
        description: A list of pets (may be filtered by type)
        schema:
          $ref: '#/pet'
        examples:
          rgb: ['name1', 'name2', 'name3']
    """
#    all_colors = {
#        'cmyk': ['cian', 'magenta', 'yellow', 'black'],
#        'rgb': ['red', 'green', 'blue']
#    }
    return jsonify({'pet': PetTable.get_all_pets()})

if __name__ == "__main__":
    app.run(port=5432, debug=True)
# swagger: http://localhost:5000/apidocs/


backend/flask/apiks_flask_API.py
#!/usr/bin/env python3

from settings import *
import json
from api import api

db = SQLAlchemy(app)

class Session_list(db.Model):
    __tablename__ = 'Session_list'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    client_id = db.Column(db.Integer, nullable=False)
    # nullable is false so the column can't be empty
    clinic_id = db.Column(db.Integer, nullable=False)
    doctor_id = db.Column(db.Integer, nullable=False)
    date_of_visit = db.Column(db.date_time_obj, nullable=False)
    price_list_id = db.Column(db.Integer, nullable=False)


# You may decorate your operation with @swagger.operation
class Session_list(db.Model):
    "get_session_list"
    @swagger.operation(
        notes='Get list of sessions',
        responseClass=ModelClass.__name__,
        nickname='session_list',
        parameters=[ ],
        responseMessages=[
            {
              "code": 200,
              "message": "List of sessions"
            },
            {
              "code": 405,
              "message": "Invalid input"
            },
            {
              "code": 503,
              "message": "Server error"
            }
          ]
        )
    def json(self):
        return {'id': self.id, 'client_id': self.clinic_id,
                'clinic_id': self.clinic_id, 'doctor_id': self.doctor_id, 'date_of_visit': self.date_of_visit, 'price_list_id': self.price_list_id}

    def add_session_list(_client_id, _clinic_id, _doctor_id, _date_of_visit, _price_list_id):
        '''function to add movie to database using _title, _year, _genre
        as parameters'''
        # creating an instance of our Movie constructor
        new_session_list = session_list(client_id=_client_id, clinic_id=_clinic_id, doctor_id=_doctor_id, date_of_visit=_date_of_visit, price_list_id=_price_list_id)
        db.session.add(new_session_list)  # add new movie to database session
        db.session.commit()  # commit changes to session

class Price_list(db.Model):
    "get price list"
    @swagger.operation(
        notes='Get price list ',
        responseClass=ModelClass.__name__,
        nickname='session_list',
        parameters=[ ],
        responseMessages=[
            {
              "code": 200,
              "message": "Price List "
            },
            {
              "code": 405,
              "message": "Invalid input"
            },
            {
              "code": 503,
              "message": "Server error"
            }
          ]
        )
    __tablename__ = 'Price_list'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    service = db.Column(db.String(30), nullable=False)
    # nullable is false so the column can't be empty
    price = db.Column(db.Integer, nullable=False)
    doctor_position_id = db.Column(db.Integer, nullable=False)

    def json(self):
        return {'id': self.id, 'service': self.service,
                'price': self.price, 'doctor_position_id': self.doctor_position_id}

class Pet_medical_card(db.Model):
    "get_session_list"
    @swagger.operation(
        notes='Get list of sessions',
        responseClass=ModelClass.__name__,
        nickname='session_list',
        parameters=[ ],
        responseMessages=[
            {
              "code": 200,
              "message": "List of sessions"
            },
            {
              "code": 405,
              "message": "Invalid input"
            },
            {
              "code": 503,
              "message": "Server error"
            }
          ]
        )
    __tablename__ = 'Pet_medical_card'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    doctor_id = db.Column(db.Integer, nullable=False)
    # nullable is false so the column can't be empty
    pet_id = db.Column(db.Integer, nullable=False)
    info = db.Column(db.String(1000), nullable=False)
    date_of_visit = db.Column(db.date_time_obj, nullable=False)

    def json(self):
        return {'id': self.id, 'doctor_id': self.doctor_id,
                'pet_id': self.pet_id, 'info': self.info, 'date_of_visit': self.date_of_visit}

class Doctor(db.Model):
    __tablename__ = 'doctor'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    name = db.Column(db.String(30), nullable=False)
    # nullable is false so the column can't be empty
    doctor_position_id = db.Column(db.Integer, nullable=False)
    phone = db.Column(db.String(13), nullable=False)
    clinic_id = db.Column(db.Integer, nullable=False)
    cabinet = db.Column(db.Integer, nullable=False)

    def json(self):
        return {'id': self.id, 'name': self.name,
                'doctor_position_id': self.doctor_position_id, 'phone': self.phone, 'clinic_id': self.clinic_id, 'cabinet': self.cabinet}

class Clinic(db.Model):
    __tablename__ = 'clinic'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    address = db.Column(db.String(50), nullable=False)
    # nullable is false so the column can't be empty
    schedule_clinic = db.Column(db.String(20), nullable=False)
    phone = db.Column(db.String(20), nullable=False)

    def json(self):
        return {'id': self.id, 'address': self.address,
                'schedule_clinic': self.schedule_clinic, 'phone': self.phone}

class Doctor_position(db.Model):
    __tablename__ = 'doctor_position'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    position = db.Column(db.String(20), nullable=False)
    # nullable is false so the column can't be empty

    def json(self):
        return {'id': self.id, 'position': self.position,}

class Pet(db.Model):
    __tablename__ = 'pet'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    pet_type_id = db.Column(db.Integer, nullable=False)
    # nullable is false so the column can't be empty
    name = db.Column(db.String(30), nullable=False)
    client_id = db.Column(db.Integer, nullable=False)
    age = db.Column(db.Integer, nullable=False)

    def json(self):
        return {'id': self.id, 'pet_type_id': self.pet_type_id,
                'name': self.name, 'client_id': self.client_id, 'age': self.age}

    def add_pet(_pet_type_id, _name, _client_id, _age):
        '''function to add movie to database using _title, _year, _genre
        as parameters'''
        # creating an instance of our Movie constructor
        new_pet = Movie(pet_type_id=_pet_type_id, name=_name, client_id=_client_id, age=_age)
        db.session.add(new_pet)  # add new movie to database session
        db.session.commit()  # commit changes to session

    def get_all_pets():
        '''function to get all movies in our database'''
        return [pet.json(pet) for pet in pet.query.all()]

    def get_pet(_name):
        '''function to get movie using the id of the movie as parameter'''
        return [pet.json(pet.query.filter_by(name=_name).first())]
        # Movie.json() coverts our output to the json format defined earlier
        # the filter_by method filters the query by the id
        # since our id is unique we will only get one result
        # the .first() method will get that first value returned

class Client(db.Model):
    __tablename__ = 'client'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    name = db.Column(db.String(30), nullable=False)
    # nullable is false so the column can't be empty
    phone = db.Column(db.String(13), nullable=False)
    email = db.Column(db.String(30), nullable=False)
    user_role_id = db.Column(db.Integer, nullable=False)

    def json(self):
        return {'id': self.id, 'name': self.name,
                'phone': self.phone, 'email': self.email, 'user_role': self.user_role}

class User_role(db.Model):
    __tablename__ = 'user_role'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    role = db.Column(db.String(15), nullable=False)
    # nullable is false so the column can't be empty

    def json(self):
        return {'id': self.id, 'role': self.role}

class Pet_type(db.Model):
    __tablename__ = 'pet_type'  # creating a table name
    id = db.Column(db.Integer, primary_key=True)  # this is the primary key
    type = db.Column(db.String(30), nullable=False)
    # nullable is false so the column can't be empty

    def json(self):
        return {'id': self.id, 'type': self.type}
